# IAL Reasoning Model — GPU Fine-Tuning (Gemma-2-2B-IT + QLoRA)

**Rewritten for GPU (Kaggle/Colab T4, 16GB).**

The original notebook targeted TPU via JAX/Flax/Tunix + GRPO (RL). That stack is heavy, TPU-only, and the notebook shows many dead-end debug cells (mock models, `TEST_MODE` simulation, RL cluster failing to construct) — it likely never completed a real training run.

Since your dataset already contains **ground-truth reasoning + answers** for every example (not just a reward signal), **supervised fine-tuning (SFT)** is the right tool here — GRPO/RL is for when you *don't* have labeled targets and need to learn from a reward signal instead. SFT will converge faster, is far more stable on a single T4, and directly teaches the model your IAL (Intention–Action–Lead) format.

**Stack:** 🤗 `transformers` + `peft` (QLoRA) + `trl` (`SFTTrainer`) + `bitsandbytes` (4-bit) — all GPU-native, no JAX/TPU dependency.

**What changed vs. the original:**
- TPU/JAX/Flax/Tunix/GRPO → GPU/PyTorch/HF Transformers/PEFT/TRL SFT
- Kaggle-secrets/kagglehub model download → `transformers` auto-download from HF Hub (gated model, needs HF token)
- No train/val/test split → proper 80/10/10 split
- No real eval → held-out test set with format-compliance rate, exact-match rate, and generation samples
- Dead mock/test-mode cells removed → real test cells added at the end


In [27]:
# Cell 1 — Install dependencies (GPU stack)
!pip install -q -U transformers accelerate peft trl bitsandbytes datasets evaluate huggingface_hub
print("Dependencies installed")


huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Dependencies installed


In [28]:
# Cell 2 — Imports & GPU check
import os, re, json, random
import numpy as np
import torch
from datasets import Dataset
from transformers import (
    AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig,
    TrainingArguments,
)
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training, PeftModel
from trl import SFTTrainer, SFTConfig

SEED = 42
random.seed(SEED); np.random.seed(SEED); torch.manual_seed(SEED)

assert torch.cuda.is_available(), "No GPU detected — enable a GPU runtime (Kaggle: Settings > Accelerator > GPU T4 x2)."
print("GPU:", torch.cuda.get_device_name(0))
print("VRAM (GB):", round(torch.cuda.get_device_properties(0).total_memory / 1e9, 1))


GPU: Tesla T4
VRAM (GB): 15.6


In [29]:
# Cell 3 — Hugging Face auth (Gemma is a gated model)
# Kaggle: add your HF token as a Kaggle Secret named "HF_TOKEN", or paste it below.
# Colab: use the Secrets panel (key icon) with the same name, or set the env var directly.
from huggingface_hub import login

HF_TOKEN = os.environ.get("HF_TOKEN", "")

try:
    from kaggle_secrets import UserSecretsClient
    if not HF_TOKEN:
        HF_TOKEN = UserSecretsClient().get_secret("HF_TOKEN")
except Exception:
    pass

if not HF_TOKEN:
    HF_TOKEN = input("Paste your Hugging Face token (needs access to google/gemma-2-2b-it): ").strip()

login(token=HF_TOKEN)
print("Logged in to Hugging Face Hub")


Paste your Hugging Face token (needs access to google/gemma-2-2b-it):  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


Logged in to Hugging Face Hub


In [30]:
# Cell 4 — Config / hyperparameters
BASE_MODEL = "google/gemma-2-2b-it"   # ~2B params, fits 16GB T4 comfortably in 4-bit
OUTPUT_DIR = "/kaggle/working/ial_gemma_lora"   # change to "./ial_gemma_lora" outside Kaggle

MAX_SEQ_LENGTH = 1024

# LoRA — reduced footprint: attention-only modules, lower rank.
# Dropping gate/up/down_proj (the big MLP matrices) is what actually shrinks trainable
# params the most, since those are much wider than the attention projections.
# If you need it even smaller/faster, drop to LORA_TARGET_MODULES = ["q_proj", "v_proj"].
LORA_R = 8
LORA_ALPHA = 16
LORA_DROPOUT = 0.05
LORA_TARGET_MODULES = ["q_proj", "k_proj", "v_proj", "o_proj"]

# Training
NUM_EPOCHS = 3
PER_DEVICE_TRAIN_BATCH_SIZE = 2
GRAD_ACCUM_STEPS = 8          # effective batch size = 2 * 8 = 16
LEARNING_RATE = 2e-4
WARMUP_RATIO = 0.05
WEIGHT_DECAY = 0.01
LOGGING_STEPS = 10
SAVE_STEPS = 20      # checkpoint frequently so a cut-off run loses at most ~20 steps
EVAL_STEPS = 20
SAVE_TOTAL_LIMIT = 3 # keep last 3 checkpoints on disk (each is small: adapter-only)


# IAL format tokens
REASONING_START, REASONING_END = "<reasoning>", "</reasoning>"
ANSWER_START, ANSWER_END = "<answer>", "</answer>"

IAL_SYSTEM_PROMPT = (
    "You are a reasoning-focused AI model.\n\n"
    "You must reason explicitly using three stages:\n"
    "1. Intention - what you are trying to accomplish\n"
    "2. Action - the operations or steps taken\n"
    "3. Lead - why these steps lead to the final answer\n\n"
    f"Respond strictly in the following format:\n\n"
    f"{REASONING_START}\n[Intention]\n...\n\n[Action]\n...\n\n[Lead]\n...\n{REASONING_END}\n\n"
    f"{ANSWER_START}\n...\n{ANSWER_END}"
)

print("Config set. Base model:", BASE_MODEL)


Config set. Base model: google/gemma-2-2b-it


In [31]:
# Cell 5 — Load & prepare the IAL dataset
DATA_PATH = "/kaggle/input/inactedformat/data (1).json"  # update this path to wherever you upload data__1_.json
if not os.path.exists(DATA_PATH):
    # fallback: look for the file the user attached in this session
    for candidate in ["data__1_.json", "/kaggle/working/data__1_.json", "data (1).json"]:
        if os.path.exists(candidate):
            DATA_PATH = candidate
            break

with open(DATA_PATH, "r") as f:
    raw_data = json.load(f)

print(f"Loaded {len(raw_data)} raw examples")
print("Keys per example:", set(raw_data[0].keys()))

def normalize_reasoning(reasoning_text: str) -> str:
    """Normalize '[Actions]' -> '[Action]' so it matches the system prompt's format."""
    return re.sub(r"\[Actions?\]", "[Action]", reasoning_text.strip())

def build_example(ex):
    instruction = ex.get("instruction", "").strip()
    input_text = ex.get("input", "").strip()
    reasoning = normalize_reasoning(ex.get("reasoning", ""))
    output = ex.get("output", "").strip()

    user_content = instruction
    if input_text:
        user_content += f"\n\nInput: {input_text}"

    target = f"{REASONING_START}\n{reasoning}\n{REASONING_END}\n\n{ANSWER_START}\n{output}\n{ANSWER_END}"

    return {
        "system": IAL_SYSTEM_PROMPT,
        "user": user_content,
        "target": target,
    }

examples = [build_example(ex) for ex in raw_data
            if ex.get("instruction") and ex.get("output") and ex.get("reasoning")]

print(f"{len(examples)} usable examples after filtering incomplete rows")
print("\n--- Sample formatted example ---")
print("USER:", examples[0]["user"][:200])
print("\nTARGET:\n", examples[0]["target"][:400])


Loaded 1383 raw examples
Keys per example: {'reasoning', 'input', 'output', 'instruction'}
1383 usable examples after filtering incomplete rows

--- Sample formatted example ---
USER: Evaluate the following phrase by transforming it into the spelling given.

Input: freind --> friend

TARGET:
 <reasoning>
[Intention]
Identify and correct orthographic error.

[Action]
Step 1: Compare input against standard spelling.
Step 2: Locate specific character sequence error.
Step 3: Provide corrected form with explanation.

[Lead]
Detection of the transposed 'i' and 'e' sequence.
</reasoning>

<answer>
The spelling of the given phrase "freind" is incorrect, the correct spelling is "friend".
</answ


In [32]:
# Cell 6 — Train / validation / test split (80 / 10 / 10)
random.shuffle(examples)
n = len(examples)
n_train = int(n * 0.8)
n_val = int(n * 0.1)

train_examples = examples[:n_train]
val_examples = examples[n_train:n_train + n_val]
test_examples = examples[n_train + n_val:]

print(f"Train: {len(train_examples)} | Val: {len(val_examples)} | Test: {len(test_examples)}")


Train: 1106 | Val: 138 | Test: 139


In [33]:
# Cell 7 — Load tokenizer, build chat-formatted text field, wrap in HF Datasets
tokenizer = AutoTokenizer.from_pretrained(BASE_MODEL, token=HF_TOKEN)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

def to_chat_text(ex):
    messages = [
        {"role": "user", "content": f"{ex['system']}\n\nInstruction:\n{ex['user']}"},
        {"role": "assistant", "content": ex["target"]},
    ]
    text = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=False)
    return {"text": text}

train_dataset = Dataset.from_list(train_examples).map(to_chat_text, remove_columns=["system", "user", "target"])
val_dataset = Dataset.from_list(val_examples).map(to_chat_text, remove_columns=["system", "user", "target"])
test_dataset = Dataset.from_list(test_examples)  # keep raw fields for generation-based testing later

print(train_dataset[0]["text"][:500])


Map:   0%|          | 0/1106 [00:00<?, ? examples/s]

Map:   0%|          | 0/138 [00:00<?, ? examples/s]

<bos><start_of_turn>user
You are a reasoning-focused AI model.

You must reason explicitly using three stages:
1. Intention - what you are trying to accomplish
2. Action - the operations or steps taken
3. Lead - why these steps lead to the final answer

Respond strictly in the following format:

<reasoning>
[Intention]
...

[Action]
...

[Lead]
...
</reasoning>

<answer>
...
</answer>

Instruction:
Describe how a rainbow is formed.<end_of_turn>
<start_of_turn>model
<reasoning>
[Intention]
Provid


In [34]:
# Cell 8 — Load base model in 4-bit (QLoRA) and attach LoRA adapters
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    bnb_4bit_use_double_quant=True,
)

model = AutoModelForCausalLM.from_pretrained(
    BASE_MODEL,
    quantization_config=bnb_config,
    device_map="auto",
    token=HF_TOKEN,
    attn_implementation="eager",  # Gemma-2 recommends eager attention
)
model.config.use_cache = False
model = prepare_model_for_kbit_training(model)

lora_config = LoraConfig(
    r=LORA_R,
    lora_alpha=LORA_ALPHA,
    lora_dropout=LORA_DROPOUT,
    target_modules=LORA_TARGET_MODULES,
    bias="none",
    task_type="CAUSAL_LM",
)

model = get_peft_model(model, lora_config)
model.print_trainable_parameters()


Loading weights:   0%|          | 0/288 [00:00<?, ?it/s]

trainable params: 3,194,880 || all params: 2,617,536,768 || trainable%: 0.1221


In [35]:
# Cell 9 — Training configuration & SFTTrainer
# Built version-safely: trl has renamed/reshuffled SFTConfig fields across releases,
# so we filter our desired args against whatever signature is actually installed.
import inspect

desired_args = dict(
    output_dir=OUTPUT_DIR,
    num_train_epochs=NUM_EPOCHS,
    per_device_train_batch_size=PER_DEVICE_TRAIN_BATCH_SIZE,
    per_device_eval_batch_size=PER_DEVICE_TRAIN_BATCH_SIZE,
    gradient_accumulation_steps=GRAD_ACCUM_STEPS,
    gradient_checkpointing=True,
    gradient_checkpointing_kwargs={"use_reentrant": False},
    learning_rate=LEARNING_RATE,
    warmup_ratio=WARMUP_RATIO,
    weight_decay=WEIGHT_DECAY,
    lr_scheduler_type="cosine",
    logging_steps=LOGGING_STEPS,
    eval_strategy="steps",       # older trl/transformers: "evaluation_strategy"
    evaluation_strategy="steps", # keep both spellings, filtered below
    eval_steps=EVAL_STEPS,
    save_strategy="steps",
    save_steps=SAVE_STEPS,
    save_total_limit=SAVE_TOTAL_LIMIT,
    bf16=True,
    optim="paged_adamw_8bit",
    max_grad_norm=0.3,
    max_length=MAX_SEQ_LENGTH,
    max_seq_length=MAX_SEQ_LENGTH,  # older trl uses this name instead of max_length
    packing=False,
    dataset_text_field="text",
    report_to="none",
    seed=SEED,
    load_best_model_at_end=True,
    metric_for_best_model="eval_loss",
)

valid_params = set(inspect.signature(SFTConfig.__init__).parameters.keys())
filtered_args = {k: v for k, v in desired_args.items() if k in valid_params}
dropped = sorted(set(desired_args) - set(filtered_args))
if dropped:
    print("Note: this trl version's SFTConfig doesn't support these args, skipping them:", dropped)

sft_config = SFTConfig(**filtered_args)

trainer = SFTTrainer(
    model=model,
    args=sft_config,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    processing_class=tokenizer,
)

print("Trainer ready. Effective batch size:",
      PER_DEVICE_TRAIN_BATCH_SIZE * GRAD_ACCUM_STEPS)


Note: this trl version's SFTConfig doesn't support these args, skipping them: ['evaluation_strategy', 'max_seq_length', 'warmup_ratio']


Adding EOS to train dataset:   0%|          | 0/1106 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/1106 [00:00<?, ? examples/s]

Building labels for train dataset:   0%|          | 0/1106 [00:00<?, ? examples/s]

Truncating train dataset:   0%|          | 0/1106 [00:00<?, ? examples/s]

Dropping fully masked examples from train dataset:   0%|          | 0/1106 [00:00<?, ? examples/s]

Adding EOS to eval dataset:   0%|          | 0/138 [00:00<?, ? examples/s]

Tokenizing eval dataset:   0%|          | 0/138 [00:00<?, ? examples/s]

Building labels for eval dataset:   0%|          | 0/138 [00:00<?, ? examples/s]

Truncating eval dataset:   0%|          | 0/138 [00:00<?, ? examples/s]

Dropping fully masked examples from eval dataset:   0%|          | 0/138 [00:00<?, ? examples/s]

Trainer ready. Effective batch size: 16


In [36]:
# Cell 10 — Train, resuming from the last checkpoint if one exists
# This means you can stop the session (timeout, disconnect, manual interrupt) and just
# re-run the whole notebook (or from Cell 8 onward) — training will pick up where it left
# off instead of starting over, as long as OUTPUT_DIR still has its checkpoint-* folders.
from transformers.trainer_utils import get_last_checkpoint

last_checkpoint = None
if os.path.isdir(OUTPUT_DIR):
    last_checkpoint = get_last_checkpoint(OUTPUT_DIR)
    if last_checkpoint:
        print(f"Found existing checkpoint, resuming from: {last_checkpoint}")
    else:
        print(f"{OUTPUT_DIR} exists but has no checkpoint yet — starting fresh")
else:
    print("No previous run found — starting fresh")

train_result = trainer.train(resume_from_checkpoint=last_checkpoint)
print(train_result.metrics)


[transformers] The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'eos_token_id': 1}.


Found existing checkpoint, resuming from: /kaggle/working/ial_gemma_lora/checkpoint-60


Step,Training Loss,Validation Loss,Entropy,Num Tokens,Mean Token Accuracy
80,0.704727,0.718880,0.343357,141232.000000,0.914403
100,0.689803,0.704113,0.338976,286709.000000,0.915885
105,0.689803,0.704097,0.338967,319969.000000,0.916044


{'train_runtime': 3192.2041, 'train_samples_per_second': 1.039, 'train_steps_per_second': 0.033, 'total_flos': 1.0381045843648512e+16, 'train_loss': 0.3045094081333705, 'epoch': 3.0}


In [37]:
# Cell 11 — Save the LoRA adapter (and tokenizer)
trainer.save_model(OUTPUT_DIR)
tokenizer.save_pretrained(OUTPUT_DIR)
print(f"Adapter + tokenizer saved to {OUTPUT_DIR}")


Adapter + tokenizer saved to /kaggle/working/ial_gemma_lora


**On resuming across sessions, not just within one:** the auto-resume in Cell 10 only helps if `OUTPUT_DIR` (`/kaggle/working/ial_gemma_lora`) still has its `checkpoint-*` folders on disk. That's true if your Kaggle session is still alive or you re-run within the same Colab runtime. It's **not** true if:
- Kaggle: you start a brand-new session (working dir resets) — commit the notebook or copy `OUTPUT_DIR` to a Kaggle Dataset first.
- Colab: your runtime disconnects (local disk is wiped) — save `OUTPUT_DIR` to Google Drive instead.

Quick Colab fix — mount Drive once and point `OUTPUT_DIR` there instead of a local path:
```python
from google.colab import drive
drive.mount('/content/drive')
OUTPUT_DIR = "/content/drive/MyDrive/ial_gemma_lora"
```
Then Cell 10's resume logic works exactly the same way, just pointed at Drive.


## Test cells

Everything below is verification: format-compliance, an exact/near-match score against held-out ground truth, and qualitative generation samples. Run after training (or point `ADAPTER_DIR` at a previously saved checkpoint to just evaluate).


In [38]:
# Test Cell 1 — Reload trained adapter for inference (fresh, memory-safe load)
import gc
del model, trainer
gc.collect()
torch.cuda.empty_cache()

ADAPTER_DIR = OUTPUT_DIR  # or point this at a saved checkpoint path

base_model_for_eval = AutoModelForCausalLM.from_pretrained(
    BASE_MODEL,
    quantization_config=bnb_config,
    device_map="auto",
    token=HF_TOKEN,
    attn_implementation="eager",
)
eval_model = PeftModel.from_pretrained(base_model_for_eval, ADAPTER_DIR)
eval_model.eval()
print("Evaluation model loaded")


Loading weights:   0%|          | 0/288 [00:00<?, ?it/s]

Evaluation model loaded


In [45]:
# Test Cell 2 — Generation helper
def generate_ial_response(instruction, input_text="", max_new_tokens=400, temperature=0.7):
    user_content = instruction
    if input_text:
        user_content += f"\n\nInput: {input_text}"
    messages = [{"role": "user", "content": f"{IAL_SYSTEM_PROMPT}\n\nInstruction:\n{user_content}"}]
    prompt = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    inputs = tokenizer(prompt, return_tensors="pt").to(eval_model.device)

    with torch.no_grad():
        output_ids = eval_model.generate(
            **inputs,
            max_new_tokens=max_new_tokens,
            temperature=temperature,
            do_sample=temperature > 0,
            top_p=0.95,
            pad_token_id=tokenizer.pad_token_id,
        )
    generated = output_ids[0][inputs["input_ids"].shape[1]:]
    return tokenizer.decode(generated, skip_special_tokens=True)

# Quick smoke test
sample_out = generate_ial_response("Calculate the sum of the first 10 positive integers.")
sample_out2 = generate_ial_response("Calculate the 5 factorial.")
print(sample_out)
print(sample_out2)


<reasoning>
[Intention]
Perform arithmetic calculation.

[Action]
Step 1: Identify arithmetic operation.
Step 2: Extract operands and coefficients.
Step 3: Apply calculation rules.

[Lead]
Recall of arithmetic series formula.
</reasoning>

<answer>
The sum of the first 10 positive integers is 55.
</answer>
<reasoning>
[Intention]
Perform mathematical computation.

[Action]
Step 1: Identify mathematical operation.
Step 2: Extract operands and calculation.
Step 3: Apply appropriate algorithm.

[Lead]
Calculation of factorial using formula.
</reasoning>

<answer>
5! = 120
</answer>


In [42]:
# Test Cell 3 — Inspect individual predictions
for r in results[:5]:
    print("INSTRUCTION:", r["instruction"])
    print("PREDICTED :", r["predicted_answer"])
    print("EXPECTED  :", r["ground_truth"])
    print("Format OK:", r["format_ok"], "| Exact match:", r["exact_match"])
    print("-" * 60)


In [43]:
# Test Cell 4 — Unit tests for the reward/format utilities (sanity checks, no GPU needed)
def _test_format_regex():
    good = f"{REASONING_START}\n[Intention]\nx\n\n[Action]\ny\n\n[Lead]\nz\n{REASONING_END}\n\n{ANSWER_START}\n42\n{ANSWER_END}"
    bad = "no tags here"
    assert check_format(good) is True, "Well-formed response should pass"
    assert check_format(bad) is False, "Malformed response should fail"
    assert extract_answer(good) == "42", "Answer extraction should pull '42'"
    print("✓ format regex tests passed")

def _test_dataset_split():
    assert len(train_examples) + len(val_examples) + len(test_examples) == len(examples)
    assert len(set(id(e) for e in train_examples) & set(id(e) for e in test_examples)) == 0
    print("✓ dataset split tests passed")

_test_format_regex()
_test_dataset_split()


✓ format regex tests passed
✓ dataset split tests passed


## Improvement ideas (beyond this rewrite)

- **More data**: 1,383 examples is small for SFT; consider augmenting with synthetic IAL-formatted examples or a public instruction-tuning dataset filtered/reformatted into this schema.
- **Curriculum**: start with `MAX_STEPS` on a small subset to sanity-check the pipeline before a full run (as the original notebook was clearly trying to do, just with mocks instead of real short runs).
- **RLHF/GRPO as a second stage**: once SFT gives you a model that reliably produces the IAL format, you *could* layer GRPO on top using the reward functions from your original notebook (format + correctness + reasoning quality) to further optimize — that's the standard SFT-then-RL recipe, and it'll be much more stable starting from an SFT checkpoint than from the raw base model.
- **Longer context**: if your `input` fields can be long, raise `MAX_SEQ_LENGTH` and watch T4 memory (drop `PER_DEVICE_TRAIN_BATCH_SIZE` to 1 if you do).
- **Try a slightly larger base model** (e.g. `Qwen2.5-3B-Instruct` or `Llama-3.2-3B-Instruct`) in 4-bit if Gemma's reasoning quality underwhelms — both fit a T4 with QLoRA using this same notebook, just swap `BASE_MODEL`.
